src/Microsoft.Graph.Core/Requests/RedirectHandler.cs

```diff
@@ -0,0 +1,172 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net.Http;
+    using System.Text;
+    using System.Net.Http.Headers;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Net;
+    
+
+    /// <summary>
+    /// An <see cref="DelegatingHandler"/> implementation using standard .NET libraries.
+    /// </summary>
+    public class RedirectHandler : DelegatingHandler
+    {
+
+        private const int maxRedirects = 5;
+
+        /// <summary>
+        /// Constructs a new <see cref="RedirectHandler"/> 
+        /// </summary>
+        /// <param name="innerHandler">An HTTP message handler to pass to the <see cref="HttpMessageHandler"/> for sending requests.</param>
+        public RedirectHandler(HttpMessageHandler innerHandler)
+        {
+            InnerHandler = innerHandler;
+        }
+
+        /// <summary>
+        /// Disposes the HttpMessageHandler instance.
+        /// </summary>
+        public new void Dispose()
+        {
+            if (InnerHandler != null)
+            {
+                InnerHandler.Dispose();
+            }
+
+        }
+
+        /// <summary>
+        /// Sends the Request 
+        /// </summary>
+        /// <param name="request">The <see cref="HttpRequestMessage"/> to send.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/>for the request.</param>
+        /// <returns>The <see cref="HttpResponseMessage"/>.</returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            
+            // send request first time to get response
+            var response = await base.SendAsync(request, cancellationToken);
+
+            // check response status code 
+            if (IsRedirect(response.StatusCode))
+            {
+
+                // Cache the original http request content 
+                StreamContent content = null;
+
+                if (request.Content != null && request.Content.Headers.ContentLength != 0)
+                {
+                    content = new StreamContent(await request.Content.ReadAsStreamAsync());
+                }
```

Comment:
I still think this entire if could be moved into CopyRequest.  The only reason I'm being picky is the next RetryHandler is going to be almost the same code as this handler so before copying it, it would be good to make all the potential changes we want.

Reply:
Yes. It would be good to make all the potential changes in the CopyRequest method based on entire logic. But I moved this out of the method is because I want to skip to copy content by ReadAsStreamAsync each time when we create a new HTTP request from the CopyRequest Method in the while loop. Just read the content once before copying. And then pass it to the method. Actually, I'm not sure which way is better? It seems a trade-off between decreasing cost and keeping entire logic.

Reply:
@darrelmiller which way do you think is better for our requirements?

Reply:
@vikyding Your gut reaction is good that you don't want to redundantly copy the request body.  However, the [StreamContent](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/StreamContent.cs) class tries really hard to be efficient and so does [HttpContent.ReadAsStreamAsync](https://github.com/dotnet/corefx/blob/a10890f4ffe0fadf090c922578ba0e606ebdd16c/src/System.Net.Http/src/System/Net/Http/HttpContent.cs).  I think it would be more valuable to encapsulate all the copying behaviour together and we can optimize for efficiency later.  We have work we will need to do when we consider streaming uploads.

Reply:
I'm 90% sure that doing new StreamContent(content.ReadAsStreamAsync()) won't actually make a new copy of the payload.